# Query Builder Data Project
___

When a report is scheduled on the BI platform, recipient(s) must be entered. The following information is stored when a report is scheduled: report name, description, subject, mail address, cc, and bcc. The information can be queried and is stored in a very messy data format in excel. The purpose of this script is to do a strategic cleaning to extract all the report information and store it into a uniform table. 

### Import Packages

In [28]:
import pandas as pd
import numpy as np
import os
import re

# change the current working directory 
# to specified path
os.chdir('C:/Users/DDR33/Software/Jupyter/Portfolio/Query Builder')

# FileName
file = 'QueryBuilder.xlsx'

# Read in the excel file
qb = pd.read_excel(file, engine='openpyxl')

# Dataframe the data will be stored in
querybuilder_data = pd.DataFrame(columns=['SI_NAME', 'SI_DESCRIPTION', 'SI_MAIL_SUBJECT', 'SI_MAIL_ADDRESSES', 'SI_MAIL_CC', 'SI_MAIL_BCC', 'SI_MAIL_MESSAGE'])

# Change how many columns are viewed
pd.set_option('display.max_columns', 1000000)
pd.set_option('display.max_rows', 1000000)

qb.head(20)

Col1                  Col2  \
0   SAP BusinessObjects Business Intelligence plat...                   NaN   
1                                                 NaN                   NaN   
2               Number of InfoObject(s) returned: 109                   NaN   
3                                                 NaN                   NaN   
4                                                 NaN                   NaN   
5                                                 NaN                   NaN   
6                                                 NaN                   NaN   
7                                               1/109                   top   
8                                          Properties                   NaN   
9                                      SI_DESCRIPTION  monthly for Teresa C   
10                                            SI_NAME  non-english x active   
11                                    Scheduling Info                   NaN   
12                                    SI_DESTINATIONS              SI_TOTAL   
13                                                NaN                     1   
14                                                NaN                   NaN   
15                                                NaN                   NaN   
16                                                NaN                   NaN   
17                                                NaN                   NaN   
18                                                NaN                   NaN   
19                                                NaN                   NaN   

                       Col3               Col4  \
0                       NaN                NaN   
1                       NaN                NaN   
2                       NaN                NaN   
3                       NaN                NaN   
4                       NaN                NaN   
5                       NaN                NaN   
6                       NaN                NaN   
7                       NaN                NaN   
8                       NaN                NaN   
9                       NaN                NaN   
10                      NaN                NaN   
11                      NaN                NaN   
12                        1                NaN   
13  SI_DEST_SCHEDULEOPTIONS     SI_SENDER_NAME   
14                      NaN    SI_MAIL_MESSAGE   
15                      NaN    SI_MAIL_SUBJECT   
16                      NaN  SI_MAIL_ADDRESSES   
17                      NaN                NaN   
18                      NaN         SI_MAIL_CC   
19                      NaN        SI_MAIL_BCC   

                                 Col5                Col6 Col7  
0                                 NaN                 NaN  NaN  
1                                 NaN                 NaN  NaN  
2                                 NaN                 NaN  NaN  
3                                 NaN                 NaN  NaN  
4                                 NaN                 NaN  NaN  
5                                 NaN                 NaN  NaN  
6                                 NaN                 NaN  NaN  
7                                 NaN                 NaN  NaN  
8                                 NaN                 NaN  NaN  
9                                 NaN                 NaN  NaN  
10                                NaN                 NaN  NaN  
11                                NaN                 NaN  NaN  
12                                NaN                 NaN  NaN  
13                                NaN                 NaN  NaN  
14  added 1/4/19 monthly for Teresa C                 NaN  NaN  
15                          %SI_NAME%                 NaN  NaN  
16                           SI_TOTAL                   1  NaN  
17                                  1  Email1@example.org  NaN  
18                           SI_TOTAL                   0  NaN  
19                           SI_TOTAL 

In [29]:
qb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2473 entries, 0 to 2472
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Col1    657 non-null    object
 1   Col2    543 non-null    object
 2   Col3    327 non-null    object
 3   Col4    1090 non-null   object
 4   Col5    1263 non-null   object
 5   Col6    936 non-null    object
 6   Col7    108 non-null    object
dtypes: object(7)
memory usage: 135.4+ KB


### Data Information
___

This is an example of the messy dataframe. The each report will be stored in the **querybuilder_data** dataframe. All the emails have been replaced with **email(number)@example.com** for the purpose of this project.

|Column|Description|
|------|-----------|
|SI_NAME|The name of the report|
|SI_DESCRIPTION|The description of what the report contains|
|SI_MAIL_SUBJECT|The subject of the email that contains the report as an attachment|
|SI_MAIL_ADDRESS|The recipient(s) of the email|
|SI_MAIL_CC|The recipient(s) of the email that are CC'd|
|SI_MAIL_BCC|The recipient(s) of the email that are BCC'd|
|SI_MAIL_MESSAGE|The body of the email|

### Clean Columns

### SI_NAME

Looking at the **qb** table in the first column(**col1**) where it is **SI_NAME**, the name of the report should be in **col2** at the same index. Each report name in **col2** will be stored in the **querybuilder_data** dataframe.

In [30]:
querybuilder_data['SI_NAME'] = qb[qb.iloc[:,0] == 'SI_NAME'].iloc[:,1].reset_index(drop=True)
querybuilder_data.head()

SI_NAME SI_DESCRIPTION  \
0                            non-english x active            NaN   
1                            CLT - Child Caseload            NaN   
2      CRCF- Client Roster by CRCF with Addresses            NaN   
3  Crosstab - Advoc & SvcAbbrev x month Dillon v2            NaN   
4              Tx - 05 Transactions by Advoc 1 v2            NaN   

  SI_MAIL_SUBJECT SI_MAIL_ADDRESSES SI_MAIL_CC SI_MAIL_BCC SI_MAIL_MESSAGE  
0             NaN               NaN        NaN         NaN             NaN  
1             NaN               NaN        NaN         NaN             NaN  
2             NaN               NaN        NaN         NaN             NaN  
3             NaN               NaN        NaN         NaN             NaN  
4             NaN               NaN        NaN         NaN             NaN

In [31]:
querybuilder_data['SI_NAME'].describe()

count                                  109
unique                                 106
top       Clt - Active Cases with 70 or 71
freq                                     2
Name: SI_NAME, dtype: object

There are 109 listed reports. So, the **querybuilder_data** dataframe should have an indexes between 0-108. Now we can process with getting the rest of the information for each report.

### SI_DESCRIPTION

The description of each report is anywhere in **col1** where **SI_DESCRIPTION** is listed, the the description will be in **col2** on the same index. 

In [32]:
querybuilder_data['SI_DESCRIPTION'] = qb[qb.iloc[:,0] == 'SI_DESCRIPTION'].iloc[:,1].reset_index(drop=True)
querybuilder_data.head()

SI_NAME  \
0                            non-english x active   
1                            CLT - Child Caseload   
2      CRCF- Client Roster by CRCF with Addresses   
3  Crosstab - Advoc & SvcAbbrev x month Dillon v2   
4              Tx - 05 Transactions by Advoc 1 v2   

                       SI_DESCRIPTION SI_MAIL_SUBJECT SI_MAIL_ADDRESSES  \
0                monthly for Teresa C             NaN               NaN   
1                   Monthly for Kathy             NaN               NaN   
2  Monthly for Christian, Becky Black             NaN               NaN   
3                  Monthly for Shawna             NaN               NaN   
4            5th of Month for Roberta             NaN               NaN   

  SI_MAIL_CC SI_MAIL_BCC SI_MAIL_MESSAGE  
0        NaN         NaN             NaN  
1        NaN         NaN             NaN  
2        NaN         NaN             NaN  
3        NaN         NaN             NaN  
4        NaN         NaN             NaN

### SI_MAIL_SUBJECT

The subject of the email containing the scheduled report is anywhere in **col4** where **SI_MAIL_SUBJECT** is listed. The subject will be on the same index in **col5**.

In [33]:
querybuilder_data['SI_MAIL_SUBJECT'] = qb[qb.iloc[:,3] == 'SI_MAIL_SUBJECT'].iloc[:,4].reset_index(drop=True)
querybuilder_data.head()

SI_NAME  \
0                            non-english x active   
1                            CLT - Child Caseload   
2      CRCF- Client Roster by CRCF with Addresses   
3  Crosstab - Advoc & SvcAbbrev x month Dillon v2   
4              Tx - 05 Transactions by Advoc 1 v2   

                       SI_DESCRIPTION  \
0                monthly for Teresa C   
1                   Monthly for Kathy   
2  Monthly for Christian, Becky Black   
3                  Monthly for Shawna   
4            5th of Month for Roberta   

                              SI_MAIL_SUBJECT SI_MAIL_ADDRESSES SI_MAIL_CC  \
0                                   %SI_NAME%               NaN        NaN   
1                    Monthly - Child Caseload               NaN        NaN   
2                      Monthly - CRCF by CRCF               NaN        NaN   
3  Monthly - xTab x Advoc & Svc abbr - Dillon               NaN        NaN   
4          5th of Month - 05 Trans by Advoc 1               NaN        NaN   

  SI_MAIL_BCC SI_MAIL_MESSAGE  
0         NaN             NaN  
1         NaN             NaN  
2         NaN             NaN  
3         NaN             NaN  
4         NaN             NaN

### RECIPIENTS
___
The recipients of the emails containing the scheduled reports is located in **col6** of the **qb** table. Since the recipient could be CC'd or BCC'd also there is a process to separate the emails. For every recipient in **col6** there is a number associated with the email at the same index in **col5**. This will help sort each email.

The **SI_TOTAL** is an important variable. A recipient could be in the main **MAIL_ADDRESS**, **MAIL_CC**, or **MAIL_BCC** category. For each of the categories the **SI_TOTAL** lists how many emails are in each category.

**EXAMPLE**  
SI_NAME: [REPORT_NAME]  
SI_DESCRIPTION: [REPORT_DESCRIPTION]  
SI_MAIL_SUBJECT: [REPORT SCHEDULE EMAIL SUBJECT]  
**SI_MAIL_ADDRESSES:** [REPORT SCHEDULE MAIN RECIPIENTS]  
* SI_TOTAL: 3 (Number of Recipients)  
        1. email  
        2. email  
        3. email

**SI_MAIL_CC** [REPORT SCHEDULE CC'd RECIPIENTS]

* SI_TOTAL: 2 (Number of Recipients)  
        1. email  
        2. email  

**SI_MAIL_BCC:** [REPORT SCHEDULE BCC'd RECIPIENTS]  
* SI_TOTAL: 1 (Number of Recipients)
        1. email  
        
The **recipient type** is another variable that will be used listed in **col4**. In order to properly sort each email by the recipient type with the correct report, the recipient type needs to be listed by each recurring recipient. The code below will add the recipient type to each email like the example below.

**Example**

|Recipient Type   |SI TOTAL|Recipient |
|-----------------|--------|----------|
|SI_MAIL_ADDRESSES|1       |email1    |
|SI_MAIL_ADDRESSES|2       |email2    |                
|SI_MAIL_ADDRESSES|3       |email3    |
|SI_MAIL_CC       |1       |email1    |
|SI_MAIL_CC       |2       |email2    |
|SI_MAIL_BCC      |1       |email1    |

In [35]:
# Filter only for recipients types and keep the null values
recipients = qb[(qb.iloc[:,3] == 'SI_MAIL_ADDRESSES') |
                    (qb.iloc[:,3] == 'SI_MAIL_CC') |
                    (qb.iloc[:,3] == 'SI_MAIL_BCC')|
                    (pd.isna(qb.iloc[:,3]))]

# List the recipient type next to each email
temp_val = ''
temp_index = 0
for i, r in recipients.iterrows():
    
    if r[3] == 'SI_MAIL_ADDRESSES':
        temp_val = r[3]
        temp_index = i
    elif (recipients.loc[temp_index][3] == temp_val) and (r[3] != 'SI_MAIL_CC') and (r[3] != 'SI_MAIL_BCC'):
        r[3] = temp_val
        
    if r[3] == 'SI_MAIL_CC':
        temp_val = r[3]
        temp_index = i
    elif (recipients.loc[temp_index][3] == temp_val) and (r[3] != 'SI_MAIL_ADDRESSES') and (r[3] != 'SI_MAIL_BCC'):
        r[3] = temp_val
        
    if r[3] == 'SI_MAIL_BCC':
        temp_val = r[3]
        temp_index = i
    elif (recipients.loc[temp_index][3] == temp_val) and (r[3] != 'SI_MAIL_ADDRESSES') and (r[3] != 'SI_MAIL_CC'):
        r[3] = temp_val

recipients = recipients.iloc[:,[3,4,5]]

# Change columns names
recipients = recipients.rename(columns={recipients.columns[0]: 'recipient_type',
                                        recipients.columns[1]: 'si_total',
                                        recipients.columns[2]: 'recipients'}).reset_index(drop=True)
recipients.head(40)

recipient_type  si_total          recipients
0                 NaN       NaN                 NaN
1                 NaN       NaN                 NaN
2                 NaN       NaN                 NaN
3                 NaN       NaN                 NaN
4                 NaN       NaN                 NaN
5                 NaN       NaN                 NaN
6                 NaN       NaN                 NaN
7                 NaN       NaN                 NaN
8                 NaN       NaN                 NaN
9                 NaN       NaN                 NaN
10                NaN       NaN                 NaN
11                NaN       NaN                 NaN
12                NaN       NaN                 NaN
13  SI_MAIL_ADDRESSES  SI_TOTAL                   1
14  SI_MAIL_ADDRESSES         1  Email1@example.org
15         SI_MAIL_CC  SI_TOTAL                   0
16        SI_MAIL_BCC  SI_TOTAL                   0
17        SI_MAIL_BCC         1        SI_MIME_TYPE
18        SI_MAIL_BCC       NaN       SI_EMBED_NAME
19        SI_MAIL_BCC       NaN                 NaN
20        SI_MAIL_BCC       NaN                 NaN
21        SI_MAIL_BCC       NaN                 NaN
22        SI_MAIL_BCC       NaN                 NaN
23        SI_MAIL_BCC       NaN                 NaN
24        SI_MAIL_BCC       NaN                 NaN
25        SI_MAIL_BCC       NaN                 NaN
26        SI_MAIL_BCC       NaN                 NaN
27  SI_MAIL_ADDRESSES  SI_TOTAL                   1
28  SI_MAIL_ADDRESSES         1  Email1@example.org
29         SI_MAIL_CC  SI_TOTAL                   0
30        SI_MAIL_BCC  SI_TOTAL                   0
31        SI_MAIL_BCC         1        SI_MIME_TYPE
32        SI_MAIL_BCC       NaN       SI_EMBED_NAME
33        SI_MAIL_BCC       NaN                 NaN
34        SI_MAIL_BCC       NaN                 NaN
35        SI_MAIL_BCC       NaN                 NaN
36        SI_MAIL_BCC       NaN                 NaN
37        SI_MAIL_BCC       NaN                 NaN
38        SI_MAIL_BCC       NaN                 NaN
39        SI_MAIL_BCC       NaN                 NaN

In the query builder, where there are multiple recipients, somtimes there are blanks where there should be an email. The code below will check if there are any missing emails.
        
The code will check the **SI_TOTAL** and if there is a missing email it will replace the null (**Nan**) value with a blank

**Example**
* SI_TOTAL: 3
    1. email
    2. email
    3. NaN

In [45]:
# Check for any possible missing emails
for i, r in recipients.iterrows():
    
    # Check SI_MAIL_ADDRESSES
    if (r['recipient_type'] == 'SI_MAIL_ADDRESSES') and (type(r['si_total']) == int) and (r['si_total'] >= 1):
        if pd.isna(r['recipients']):
            r['recipients'] = ''
            
    # Check SI_MAIL_CC
    if (r['recipient_type'] == 'SI_MAIL_CC') and (type(r['si_total']) == int) and (r['si_total'] >= 1):
        if pd.isna(r['recipients']):
            r['recipients'] = ''
            
    # Check SI_MAIL_BCC
    if (r['recipient_type'] == 'SI_MAIL_BCC') and (type(r['si_total']) == int) and (r['si_total'] >= 1):
        if pd.isna(r['recipients']):
            r['recipients'] = ''
                
# Filter the the column to contain only emails and possible missing emails
recipients = recipients[(recipients['recipients'].str.contains('@', flags=re.I, na=False, regex=True)) |
                        (recipients['recipients'] == '')].reset_index(drop=True)
recipients.iloc[49:62]

recipient_type si_total          recipients
49  SI_MAIL_ADDRESSES        1  Email1@example.org
50  SI_MAIL_ADDRESSES        1  Email1@example.org
51  SI_MAIL_ADDRESSES        1  Email1@example.org
52  SI_MAIL_ADDRESSES        1  Email1@example.org
53  SI_MAIL_ADDRESSES        2  Email2@example.org
54  SI_MAIL_ADDRESSES        1  Email1@example.org
55  SI_MAIL_ADDRESSES        2  Email2@example.org
56         SI_MAIL_CC        1  Email1@example.org
57  SI_MAIL_ADDRESSES        1  Email1@example.org
58  SI_MAIL_ADDRESSES        2  Email2@example.org
59  SI_MAIL_ADDRESSES        3  Email3@example.org
60         SI_MAIL_CC        1  Email1@example.org
61         SI_MAIL_CC        2  Email2@example.org

Now that each recipient type is next to each recipient the emails can now be sorted to each report. The logic behind this is every email has to have recipients in the **SI_MAIL_ADDRESSES** category.

**Example:**

|Subject:|Email Subject|
|---|---------|
|To:|email1@example.com, email2@example.com, email3@example.com|

This example would look like this in the **recipients table**

|Recipient Type   |SI_TOTAL|Recipient |
|-----------------|--------|----------|
|SI_MAIL_ADDRESSES|1       |email1@example.com    |
|SI_MAIL_ADDRESSES|2       |email2@example.com    |                
|SI_MAIL_ADDRESSES|3       |email3@example.com    |

Sometimes there will be recipients that are CC'd or BCC'd

**Example:**

|Subject:|Email Subject|
|---|---------|
|To:|email1@example.com, email2@example.com, email3@example.com|
|CC:|email1@example.com, email2@example.com|
|BCC:|email1@example.com, email2@example.com|

This example would look like this in the **recipients table**


|Recipient Type   |SI_TOTAL|Recipient |
|-----------------|--------|----------|
|SI_MAIL_ADDRESSES|1       |email1@example.com    |
|SI_MAIL_ADDRESSES|2       |email2@example.com    |                
|SI_MAIL_ADDRESSES|3       |email3@example.com    |
|SI_MAIL_CC       |1       |email1@example.com    |
|SI_MAIL_CC       |2       |email2@example.com    |
|SI_MAIL_BCC       |1       |email1@example.com    |
|SI_MAIL_BCC       |2       |email2@example.com    |

If a recipient in the **SI_MAIL_CC** or/and **SI_MAIL_BCC** category follows the **SI_MAIL_ADDRESSES** category, it will be sorted into the same email. The code below will sort each email into its respective category.

In [50]:
# index counters for each recipient type
recipient_index = 0

for i, r in recipients.iterrows():
    
    if r['si_total'] == 1:
        if r['recipient_type'] == 'SI_MAIL_ADDRESSES':
            querybuilder_data.at[recipient_index, 'SI_MAIL_ADDRESSES'] = r['recipients']
            recipient_index += 1
        if r['recipient_type'] == 'SI_MAIL_CC':
            querybuilder_data.at[recipient_index - 1, 'SI_MAIL_CC'] = r['recipients']
        if r['recipient_type'] == 'SI_MAIL_BCC':
            querybuilder_data.at[recipient_index - 1, 'SI_MAIL_BCC'] = r['recipients']
    else:
        if r['recipient_type'] == 'SI_MAIL_ADDRESSES':
            querybuilder_data.at[recipient_index - 1, 'SI_MAIL_ADDRESSES'] = querybuilder_data.at[recipient_index - 1, 'SI_MAIL_ADDRESSES'] + ', ' + r['recipients']
        if r['recipient_type'] == 'SI_MAIL_CC':
            querybuilder_data.at[recipient_index - 1, 'SI_MAIL_CC'] = querybuilder_data.at[recipient_index - 1, 'SI_MAIL_CC'] + ', ' + r['recipients']
        if r['recipient_type'] == 'SI_MAIL_BCC':
            querybuilder_data.at[recipient_index - 1, 'SI_MAIL_BCC'] = querybuilder_data.at[recipient_index - 1, 'SI_MAIL_BCC'] + ', ' + r['recipients']
            
querybuilder_data.iloc[27:39]

SI_NAME  \
27             620 Tx- Cancel NoShow Detail shows srv   
28             603 Tx- Cancel NoShow Detail shows srv   
29             560 Tx- Cancel NoShow Detail shows srv   
30             539 Tx- Cancel NoShow Detail shows srv   
31             376 Tx- Cancel NoShow Detail shows srv   
32  EOC - Self-Pay Balances for Clients on DHEC De...   
33                          noshow cfield new 7-19-19   
34              Tx-Direct Services x Locat x Serv1 v2   
35              640 Tx- Cancel NoShow Detail w srv v3   
36             709 Tx- Cancel NoShow Detail shows srv   
37                                     staff registry   
38                   staff registry with non-billable   

                               SI_DESCRIPTION  \
27                    weekly for Evonne Young   
28                   weekly for Cindy Griffin   
29                    weekly for Ridgill Mims   
30              weekly for Priscilla Morrison   
31    automatically generated for Becky Black   
32                                        NaN   
33  new no show no onit daily for front staff   
34                  Monthly for Christian, OM   
35                       weekly for Anna Paul   
36                 weekly for Miranda Roberts   
37                       monthly for Teresa C   
38                       monthly for Teresa C   

                           SI_MAIL_SUBJECT  \
27                               %SI_NAME%   
28                               %SI_NAME%   
29                               %SI_NAME%   
30                               %SI_NAME%   
31                               %SI_NAME%   
32             monthly - DHEC death report   
33                               %SI_NAME%   
34  Monthly - Tx Direct Serv x Loc x Serv1   
35                               %SI_NAME%   
36                               %SI_NAME%   
37                               %SI_NAME%   
38                               %SI_NAME%   

                                    SI_MAIL_ADDRESSES  \
27                                 Email1@example.org   
28                                 Email1@example.org   
29                                 Email1@example.org   
30                                 Email1@example.org   
31                                 Email1@example.org   
32             Email1@example.org, Email2@example.org   
33             Email1@example.org, Email2@example.org   
34  Email1@example.org, Email2@example.org, Email3...   
35                                 Email1@example.org   
36                                 Email1@example.org   
37                                 Email1@example.org   
38                                 Email1@example.org   

                                SI_MAIL_CC         SI_MAIL_BCC SI_MAIL_MESSAGE  
27                                     NaN                 NaN             NaN  
28                                     NaN                 NaN             NaN  
29                                     NaN                 NaN             NaN  
30                                     NaN                 NaN             NaN  
31                                     NaN                 NaN             NaN  
32                                     NaN                 NaN             NaN  
33                      Email1@example.org                 NaN             NaN  
34  Email1@example.org, Email2@example.org                 NaN             NaN  
35                                     NaN                 NaN             NaN  
36                                     NaN                 NaN             NaN  
37                                     NaN  Email1@example.org             NaN  
38                                     NaN  Email1@example.org             NaN

In [52]:
print('Number of rows: ', querybuilder_data.shape[0])

Number of rows:  109


##### There are still 109 rows corresponding to the 109 different reports that were counted at the beginning of the project. This means that the emails have been sorted correctly.

### SI_MAIL_MESSAGE

Now that the recipients are sorted by the correct reports the body of each email now can be sorted. The body of the email is anywhere in **col4** where **SI_MAIL_MESSAGE** is listed. The body will be at the same index in **col5**.

In [53]:
querybuilder_data['SI_MAIL_MESSAGE'] = qb[qb.iloc[:,3] == 'SI_MAIL_MESSAGE'].iloc[:,4].reset_index(drop=True)
querybuilder_data.head(10)

SI_NAME  \
0                            non-english x active   
1                            CLT - Child Caseload   
2      CRCF- Client Roster by CRCF with Addresses   
3  Crosstab - Advoc & SvcAbbrev x month Dillon v2   
4              Tx - 05 Transactions by Advoc 1 v2   
5                   Tx- H056 H057 H058 H059 FY 18   
6              MCO in H003, 004, 005 w names FY19   
7                Clt - FollowUp Consent Labels v2   
8              Clt - Followup Consent Y's only v3   
9                 Clt - Active Clients w SC Prime   

                       SI_DESCRIPTION  \
0                monthly for Teresa C   
1                   Monthly for Kathy   
2  Monthly for Christian, Becky Black   
3                  Monthly for Shawna   
4            5th of Month for Roberta   
5      weekly on Fridays for Teresa C   
6      weekly on Fridays for Teresa C   
7                  Monthly for Kaylah   
8                  Monthly for Kaylah   
9                 Monthly for Roberta   

                              SI_MAIL_SUBJECT  \
0                                   %SI_NAME%   
1                    Monthly - Child Caseload   
2                      Monthly - CRCF by CRCF   
3  Monthly - xTab x Advoc & Svc abbr - Dillon   
4          5th of Month - 05 Trans by Advoc 1   
5                                   %SI_NAME%   
6                                   %SI_NAME%   
7      Monthly - Clt Follow Up Consent Labels   
8                  Monthly - Followup Consent   
9                monthly - Act Cli w SC Prime   

                                   SI_MAIL_ADDRESSES SI_MAIL_CC SI_MAIL_BCC  \
0                                 Email1@example.org        NaN         NaN   
1                                 Email1@example.org        NaN         NaN   
2             Email1@example.org, Email2@example.org        NaN         NaN   
3                                 Email1@example.org        NaN         NaN   
4                                 Email1@example.org        NaN         NaN   
5                                 Email1@example.org        NaN         NaN   
6                                 Email1@example.org        NaN         NaN   
7                                 Email1@example.org        NaN         NaN   
8                                 Email1@example.org        NaN         NaN   
9  Email1@example.org, Email2@example.org, Email3...        NaN         NaN   

                                     SI_MAIL_MESSAGE  
0                  added 1/4/19 monthly for Teresa C  
1                            Automatically generated  
2                            Automatically generated  
3                            Automatically generated  
4  automatically generated updated 7/11/18 to rep...  
5                   fixed migration schedule 3/18/19  
6                   fixed migration schedule 3/18/19  
7                            Automatically generated  
8                            Automatically generated  
9  automatically generated updated 7/11/18 to rep...

# Final Results

In [54]:
querybuilder_data

SI_NAME  \
0                                 non-english x active   
1                                 CLT - Child Caseload   
2           CRCF- Client Roster by CRCF with Addresses   
3       Crosstab - Advoc & SvcAbbrev x month Dillon v2   
4                   Tx - 05 Transactions by Advoc 1 v2   
5                        Tx- H056 H057 H058 H059 FY 18   
6                   MCO in H003, 004, 005 w names FY19   
7                     Clt - FollowUp Consent Labels v2   
8                   Clt - Followup Consent Y's only v3   
9                      Clt - Active Clients w SC Prime   
10                    Clt - Active Cases with 70 or 71   
11                    Clt - Active Cases with 70 or 71   
12             Clt - Discharged Clients by Location v2   
13           CLT - No PMA master list- w Mcaid numbers   
14                  CRCF - Client Alpha Roster by CSMG   
15                      Clt- Mcare#s with ABN dates v2   
16               Tx- Cancel NoShow Detail shows srv v2   
17               579 Tx- Cancel NoShow Detail w srv v3   
18        Crosstab - Advoc & SvcAbbrev x month Ches v2   
19              696 Tx- Cancel NoShow Detail shows srv   
20              687 Tx- Cancel NoShow Detail shows srv   
21              679 Tx- Cancel NoShow Detail shows srv   
22              675 Tx- Cancel NoShow Detail shows srv   
23              665 Tx- Cancel NoShow Detail shows srv   
24              650 Tx- Cancel NoShow Detail shows srv   
25              651 Tx- Cancel NoShow Detail shows srv   
26              623 Tx- Cancel NoShow Detail shows srv   
27              620 Tx- Cancel NoShow Detail shows srv   
28              603 Tx- Cancel NoShow Detail shows srv   
29              560 Tx- Cancel NoShow Detail shows srv   
30              539 Tx- Cancel NoShow Detail shows srv   
31              376 Tx- Cancel NoShow Detail shows srv   
32   EOC - Self-Pay Balances for Clients on DHEC De...   
33                           noshow cfield new 7-19-19   
34               Tx-Direct Services x Locat x Serv1 v2   
35               640 Tx- Cancel NoShow Detail w srv v3   
36              709 Tx- Cancel NoShow Detail shows srv   
37                                      staff registry   
38                    staff registry with non-billable   
39    Contacts vs Undup Contacts x Srv x age x prev mo   
40                             694 Notes Signed Status   
41                             701 Notes Signed Status   
42                    Tx- Cancel NoShow for doctors v3   
43              710 Tx- Cancel NoShow Detail shows srv   
44                    Clt - Financial Eval by Locat Q1   
45                    Clt - Financial Eval by Locat Q2   
46                    Clt - Financial Eval by Locat Q3   
47                    Clt - Financial Eval by Locat Q4   
48                      CLT - MCAID active by location   
49                   CLT - NO MCAID active by location   
50                            Clt- MCare Admissions v2   
51   Tx- Clt Missing Srvcs H003 H004 H005 not admit...   
52         Clt - Active Caseload with Addresses by Loc   
53                    Lookup-Insurance Code Listing v2   
54              Clt - Last Service 45 days v2 - Bville   
55             Clt - Last Service 45 days v2 - C'field   
56              Clt - Last Service 45 days v2 - Dillon   
57                             586 Notes Signed Status   
58                                   Self Pay Bills v5   
59    CLT - Child Caseload for Voter Registration Form   
60                      Clt - missing clinical info v2   
61   Engagement 1.2 - Caseload x CSMG - wi last SRV...   
62                        Adm - Bed Days x 10th of Mth   
63         Adm - Hosp Admissions for the last month v2   
64           Clt - Active Caseload - Last Srv1 no H052   
65               Clt - Active Caseload - Last Srv1 PMA   
66                  Clt - Active Caseload - Summary v2   
67   Clt - Active Cases to Review for Closing - No ...   
68   Clt - Caseload Report x Case Manager x C

### Store DataFrame into Excel

In [55]:
querybuilder_data.to_excel('QueryBuilder_cleaned.xlsx', index=False)